php-checks/src/main/java/org/sonar/php/checks/security/POSIXFilePermissionsCheck.java

```diff
@@ -0,0 +1,112 @@
+import org.sonar.php.tree.visitors.AssignmentExpressionVisitor;
+import org.sonar.plugins.php.api.symbols.Symbol;
+import org.sonar.plugins.php.api.tree.CompilationUnitTree;
+import org.sonar.plugins.php.api.tree.SeparatedList;
+import org.sonar.plugins.php.api.tree.Tree.Kind;
+import org.sonar.plugins.php.api.tree.declaration.CallArgumentTree;
+import org.sonar.plugins.php.api.tree.expression.ExpressionTree;
+import org.sonar.plugins.php.api.tree.expression.FunctionCallTree;
+import org.sonar.plugins.php.api.tree.expression.MemberAccessTree;
+import org.sonar.plugins.php.api.visitors.PHPVisitorCheck;
+
+@Rule(key = "S2612")
+public class POSIXFilePermissionsCheck extends PHPVisitorCheck {
+
+  private static final String MESSAGE = "Make sure this permission is safe.";
+
+  private AssignmentExpressionVisitor assignmentExpressionVisitor;
+
+  @Override
+  public void visitCompilationUnit(CompilationUnitTree tree) {
+    assignmentExpressionVisitor = new AssignmentExpressionVisitor(context().symbolTable());
+    tree.accept(assignmentExpressionVisitor);
+    super.visitCompilationUnit(tree);
+  }
+
+  @Override
+  public void visitFunctionCall(FunctionCallTree tree) {
+    ExpressionTree callee = tree.callee();
+    if (callee.is(Kind.OBJECT_MEMBER_ACCESS)) {
+      MemberAccessTree chmodAccessTree = (MemberAccessTree) callee;
+      if (chmodAccessTree.member().toString().equals("chmod")) {
+        chmodCheck(tree, tree.callArguments());
+      }
+    }
+    if (callee.toString().equals("chmod")) {
+      chmodCheck(tree, tree.callArguments());
+    }
+    if (callee.toString().equals("umask")) {
+      umaskCheck(tree, tree.callArguments());
+    }
+    super.visitFunctionCall(tree);
+  }
+
+  private void chmodCheck(FunctionCallTree tree, SeparatedList<CallArgumentTree> arguments) {
+    if (arguments.size() >= 2) {
+      int mode = resolveArgument(arguments.get(1), 0);
+      int umask = arguments.size() >= 3 ? resolveArgument(arguments.get(2), 0) : 0;
+
+      if ((mode & ~umask) % 8 != 0) {
+        context().newIssue(this, tree, MESSAGE);
+      }
+    }
+  }
+
+  private void umaskCheck(FunctionCallTree tree, SeparatedList<CallArgumentTree> arguments) {
+    if (!arguments.isEmpty()) {
+      int mask = resolveArgument(arguments.get(0), 7);
+      if (mask % 8 != 7) {
+        context().newIssue(this, tree, MESSAGE);
+      }
+    }
+  }
+
+  private int resolveArgument(CallArgumentTree modeArgument, int defaultValue) {
+    if (modeArgument.value().is(Kind.VARIABLE_IDENTIFIER)) {
+      Symbol symbol = context().symbolTable().getSymbol(modeArgument.value());
+      Optional<ExpressionTree> uniqueAssignedValue = assignmentExpressionVisitor.getUniqueAssignedValue(symbol);
+      if (uniqueAssignedValue.isPresent()) {
+        return getDecimalRepresentation(uniqueAssignedValue.get().toString(), defaultValue);
+      }
+    }
+    return getDecimalRepresentation(modeArgument.value().toString(), defaultValue);
```

Comment:
As a general rule, using `PHPTree.toString` should be discouraged. It requires going through all tokens which can be expensive for large expressions. An alternative is to test the tree kind, cast to the appropriate interface, and call one of its methods. It's more cumbersome but also more explicit. We may define utility methods to make it easier.

Reply:
Thanks for explaining, I fixed it trying to follow your suggestion.